dot/sync/chain_sync.go

```diff
@@ -0,0 +1,785 @@
+			for _, bd := range resp.BlockData[i:] {
+				body, err := types.NewBodyFromOptional(bd.Body)
+				if err != nil {
+					return fmt.Errorf("failed to convert block body from optional: hash=%s err=%s", bd.Hash, err)
+				}
+
+				cs.pendingBlocks.addBlock(&types.Block{
+					Header: curr,
+					Body:   body,
+				})
+			}
+			return errResponseIsNotChain
+		}
+
+		prev = curr
+	}
+
+	return nil
+}
+
+// validateBlockData checks that the expected fields are in the block data
+func validateBlockData(req *network.BlockRequestMessage, bd *types.BlockData) error {
+	if bd == nil {
+		return errNilBlockData
+	}
+
+	requestedData := req.RequestedData
+
+	if (requestedData&network.RequestedDataHeader) == 1 && bd.Header == nil {
+		return errNilHeaderInResponse
+	}
+
+	if (requestedData&network.RequestedDataBody>>1) == 1 && bd.Body == nil {
+		return errNilBodyInResponse
+	}
+
+	return nil
+}
+
+func workerToRequests(w *worker) ([]*network.BlockRequestMessage, error) {
+	// worker must specify a start number
+	// empty start hash is ok (eg. in the case of bootstrap, start hash is unknown)
+	if w.startNumber == nil {
+		return nil, errWorkerMissingStartNumber
+	}
+
+	// worker must specify a target number
+	// empty target hash is ok (eg. in the case of descending fork requests)
+	if w.targetNumber == nil {
+		return nil, errWorkerMissingTargetNumber
+	}
+
+	diff := big.NewInt(0).Sub(w.targetNumber, w.startNumber)
+	if diff.Int64() < 0 && w.direction != network.Descending {
+		return nil, errInvalidDirection
+	}
+
+	if diff.Int64() > 0 && w.direction != network.Ascending {
+		return nil, errInvalidDirection
+	}
+
+	// to deal with descending requests (ie. target may be lower than start) which are used in tip mode,
+	// take absolute value of difference between start and target
+	numBlocks := int(big.NewInt(0).Abs(diff).Int64())
+	numRequests := numBlocks / maxResponseSize
+
+	if numBlocks%maxResponseSize != 0 {
+		numRequests++
+	}
+
+	startNumber := w.startNumber.Uint64()
+	reqs := make([]*network.BlockRequestMessage, numRequests)
+
+	for i := 0; i < numRequests; i++ {
+		// check if we want to specify a size
+		var max *optional.Uint32
+		if i == numRequests-1 {
+			size := numBlocks % maxResponseSize
+			if size == 0 {
+				size = maxResponseSize
+			}
+			max = optional.NewUint32(true, uint32(size))
+		} else {
+			max = optional.NewUint32(false, 0)
+		}
+
+		var start *variadic.Uint64OrHash
+		if w.startHash.Equal(common.EmptyHash) {
+			// worker startHash is unspecified if we are in bootstrap mode
+			start, _ = variadic.NewUint64OrHash(startNumber)
+		} else {
+			// in tip-syncing mode, we know the hash of the block on the fork we wish to sync
+			start, _ = variadic.NewUint64OrHash(w.startHash)
+		}
+
+		var end *optional.Hash
+		if w.targetHash.Equal(common.EmptyHash) {
+			end = optional.NewHash(false, common.Hash{})
+		} else {
+			end = optional.NewHash(true, w.targetHash)
+		}
+
+		reqs[i] = &network.BlockRequestMessage{
```

Comment:
the maximum response size is hard coded to 128 (in gossamer and substrate), so even if the block range is larger, it will only go up to 128. I could also leave the target empty for bootstrap, but for tip we need to specify the target